### PR TITLE
audit fixes

### PR DESCRIPTION
**Pull request by alyn509**
fixes based on audit:

https://docs.google.com/document/d/1Tye-zOG6z6vAI_bajvU5iHjlMzel3K8uXZK7sq-adYk/edit

### CODE REVIEW

File: liquid-staking/src/lib.rs [hunk 1]
```diff
@@ -227,13 +225,59 @@ pub trait LiquidStaking<ContractReader>:
...
+                continue;
+            }
+
+            delegation_contract_mapper.update(|contract_data| {
+                contract_data.total_unstaked_from_ls_contract -= &unstake_amount;
+                contract_data.total_unbonded_from_ls_contract -= &unstake_amount
+            });
+
+            total_unstake_amount += unstake_amount;
+            providers_unbond_from.push(unstake_token_attributes);
```
Comment (psorinionut) [inline liquid-staking/src/lib.rs:265]: What's the use of this variable?

File: liquid-staking/src/lib.rs [hunk 2]
```diff
@@ -255,78 +299,49 @@ pub trait LiquidStaking<ContractReader>:
...
         #[call_result] result: ManagedAsyncCallResult<()>,
     ) {
         match result {
             ManagedAsyncCallResult::Ok(()) => {
                 let withdraw_amount = self.call_value().egld_value().clone_value();
                 let mut storage_cache = StorageCache::new(self);
                 let delegation_contract_mapper =
-                    self.delegation_contract_data(&delegation_contract);
+                    self.delegation_contract_data(&unstake_token_attributes.delegation_contract);
                 if withdraw_amount > 0u64 {
```
Comment (psorinionut) [inline liquid-staking/src/lib.rs:334]: It is important to keep in mind that the withdraw process happens at the SC's level, and not for a specific user. I think we should only update the storage (total_withdrawn_egld and total_unbonded_from_ls_contract) inside the if statement withdraw_amount > 0u64. In other words, we should update unbond_from_provider and burn the tokens, regardless the withdraw amount (as the tokens may have been previously withdrawn by other user).
Moreover, maybe we would be more safe if we cover the case of an error as well (by sending back the tokens to the user if the withdraw callback receives an error).

File: liquid-staking/src/lib.rs [hunk 3]
```diff
@@ -227,13 +225,59 @@ pub trait LiquidStaking<ContractReader>:
...
+                continue;
+            }
+
+            let delegation_contract = unstake_token_attributes.delegation_contract.clone();
+            let unstake_amount = unstake_token_attributes.unstake_amount.clone();
+            let delegation_contract_mapper = self.delegation_contract_data(&delegation_contract);
+            let delegation_contract_data = delegation_contract_mapper.get();
+
+            if delegation_contract_data.total_unbonded_from_ls_contract < unstake_amount {
+                continue;
```
Comment (psorinionut) [inline liquid-staking/src/lib.rs:256]: We should really not get to this point, as all unbonds are correlated with an actual user. I would keep the check, but instead panic the SC if it gets to this point.

File: liquid-staking/src/lib.rs [hunk 4]
```diff
@@ -227,13 +225,59 @@ pub trait LiquidStaking<ContractReader>:
...
+        }
+
+        storage_cache.total_withdrawn_egld -= &total_unstake_amount;
+        self.unstake_token_supply()
+            .update(|x| *x -= &total_unstake_amount);
+        self.send().direct_egld(&caller, &total_unstake_amount)
+    }
+
+    #[payable("*")]
+    #[endpoint(withdrawAll)]
```
Comment (psorinionut) [inline liquid-staking/src/lib.rs:275]: Not sure of the naming, as we only withdraw from a single provider, not from "all".

**Review by psorinionut — CHANGES_REQUESTED** [review]

File: liquid-staking/src/lib.rs (hunk 4 above)
Comment (alyn509) [reply liquid-staking/src/lib.rs:279]: Jacob's idea

File: liquid-staking/src/lib.rs (hunk 1 above)
Comment (alyn509) [reply liquid-staking/src/lib.rs]: not sure which one you refer to.. anyway the logic was changed

File: liquid-staking/src/delegation.rs [hunk 5]
```diff
@@ -285,6 +280,20 @@ pub trait DelegationModule:
                 .get_sc_balance(&EgldOrEsdtTokenIdentifier::egld(), 0)
                 - current_total_withdrawn_egld;
         }
+
+        let mut last_node = current_claim_status.current_node;
+        let mut delegation_addresses: ManagedVec<Self::Api, ManagedAddress> = ManagedVec::new();
```
[the remark this answers is not in the document]
Comment (alyn509) [reply liquid-staking/src/delegation.rs]: changed to a mapper. This should iterate over and pop addresser from after claiming, as it was suggested in the audit

File: .github/workflows/actions.yml [hunk 6]
```diff
@@ -8,15 +8,14 @@ on:
...
 
 jobs:
   contracts:
     name: Contracts
-    uses: multiversx/mx-sc-actions/.github/workflows/contracts.yml@v2.3.0
+    uses: multiversx/mx-sc-actions/.github/workflows/contracts.yml@v2.3.5
     with:
-      rust-toolchain: nightly-2023-04-24
-      vmtools-version: v1.4.60
+      rust-toolchain: nightly-2023-12-11
```
Comment (andrei-marinica) [inline .github/workflows/actions.yml:18]: stable

Comment (alyn509) [reply .github/workflows/actions.yml:18]: this didn't solve anything

**Review by andrei-marinica — APPROVED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Pull request merged**